jhu-idc/idc-isle-buildkit
# Add support for per-microservice JMS tuning.

This PR enables the following JMS properties to be varied on a per-microservice basis:
- maximum number of connections between the microservice and the broker (this is, in effect, the size of the connection pool)
- number of consumers reading from the microservice's queue on the broker
- acknowledgement mode used by the microservice when reading messages from the queue
- broker url and credentials

Default values can be configured and overridden on a per-microservice basis.

> Notably, this PR updates the default acknowledgement mode from `AUTO_ACKNOWLEDGE` to `CLIENT_ACKNOWLEDGE` for all microservices!

Prior to this PR all the microservices shared a single broker instance which was published via JNDI (via `fcrepo-service-activemq` from `fcrepo-camel-toolbox`).  This PR instantiates a broker for each microservice which allows for individual tuning of JMS parameters, accomplished by instantiating the necessary components in each microservice's blueprint xml.

### To Test

1. Check out and build the alpaca image for this PR (`./gradlew -Plocal alpaca:build`)
2. The general pattern is to start alpaca, set any environment variables, and view the blueprint xml files that are produced by `confd`.

View default values of JMS properties; they should align with documented defaults:
```shell
docker run --rm \
  local/alpaca bash -lc \
  'for i in fits homarus houdini ocr ; do \
      echo ">>> deploy/ca.islandora.alpaca.connector.$i.blueprint.xml" ; \
      egrep name=\"jms\..*\"  deploy/ca.islandora.alpaca.connector.$i.blueprint.xml ; \
  done'

```

Set default JMS properties across all microservices; all values for all microservices should use the provided default values:
```shell
docker run --rm \
  -e ALPACA_ACTIVEMQ_URL=tcp://somethingelse:61616 \
  -e ALPACA_ACTIVEMQ_USER=someuser \
  -e ALPACA_ACTIVEMQ_PASSWORD=somepassword \
  -e ALPACA_ACTIVEMQ_JMS_MAXCONNECTIONS=25 \
  -e ALPACA_ACTIVEMQ_JMS_CONSUMERS=10 \
  -e ALPACA_ACTIVEMQ_JMS_ACKNOWLEDGEMENT_MODE=AUTO_ACKNOWLEDGE \
  local/alpaca bash -lc \
  'for i in fits homarus houdini ocr ; do \
      echo ">>> deploy/ca.islandora.alpaca.connector.$i.blueprint.xml" ; \
      egrep name=\"jms\..*\"  deploy/ca.islandora.alpaca.connector.$i.blueprint.xml ; \
  done'
```

Override the set defaults for the Houdini container; all services except Houdini use the provided defaults:
```shell
docker run --rm   \
  -e ALPACA_ACTIVEMQ_URL=tcp://somethingelse:61616 \
  -e ALPACA_ACTIVEMQ_USER=someuser \
  -e ALPACA_ACTIVEMQ_PASSWORD=somepassword \
  -e ALPACA_ACTIVEMQ_JMS_MAXCONNECTIONS=25 \
  -e ALPACA_ACTIVEMQ_JMS_CONSUMERS=10 \
  -e ALPACA_ACTIVEMQ_JMS_ACKNOWLEDGEMENT_MODE=AUTO_ACKNOWLEDGE \
  -e ALPACA_HOUDINI_ACTIVEMQ_URL=tcp://houdinibroker:61616 \
  -e ALPACA_HOUDINI_ACTIVEMQ_USER=houdiniuser \
  -e ALPACA_HOUDINI_ACTIVEMQ_PASSWORD=houdinipassword \
  -e ALPACA_HOUDINI_ACTIVEMQ_JMS_MAXCONNECTIONS=200 \
  -e ALPACA_HOUDINI_ACTIVEMQ_JMS_CONSUMERS=15 \
  -e ALPACA_HOUDINI_ACTIVEMQ_JMS_ACKNOWLEDGEMENT_MODE=DUPS_OK_ACKNOWLEDGE \
  local/alpaca bash -lc \
  'for i in fits homarus houdini ocr ; do \
      echo ">>> deploy/ca.islandora.alpaca.connector.$i.blueprint.xml" ; \
      egrep name=\"jms\..*\"  deploy/ca.islandora.alpaca.connector.$i.blueprint.xml ; \
  done'
```

Override defaults for OCR container; all services except OCR should used documented defaults:
```shell
docker run --rm   \
  -e ALPACA_OCR_ACTIVEMQ_URL=tcp://ocrbroker:61616 \
  -e ALPACA_OCR_ACTIVEMQ_USER=ocruser \
  -e ALPACA_OCR_ACTIVEMQ_PASSWORD=ocrpassword \
  -e ALPACA_OCR_ACTIVEMQ_JMS_MAXCONNECTIONS=100 \
  -e ALPACA_OCR_ACTIVEMQ_JMS_CONSUMERS=35 \
  -e ALPACA_OCR_ACTIVEMQ_JMS_ACKNOWLEDGEMENT_MODE=DUPS_OK_ACKNOWLEDGE \
  local/alpaca bash -lc \
  'for i in fits homarus houdini ocr ; do \
      echo ">>> deploy/ca.islandora.alpaca.connector.$i.blueprint.xml" ; \
      egrep name=\"jms\..*\"  deploy/ca.islandora.alpaca.connector.$i.blueprint.xml ; \
  done'
```
